programs/local/LocalServer.cpp

```diff
@@ -191,49 +178,315 @@ static void attachSystemTables(ContextPtr context)
+    }
+    catch (...)
+    {
+        if (!ignore_error)
+            throw;
+
+        local_server_exception = std::make_unique<Exception>(getCurrentExceptionMessage(true), getCurrentExceptionCode());
+        have_error = true;
+    }
+}
+
+
+static ConfigurationPtr getConfigurationFromXMLString(const char * xml_data)
+{
+    std::stringstream ss{std::string{xml_data}};    // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    Poco::XML::InputSource input_source{ss};
+    return {new Poco::Util::XMLConfiguration{&input_source}};
+}
+
+
+void LocalServer::setupUsers()
+{
+    static const char * minimal_default_user_xml =
+        "<yandex>"
+        "    <profiles>"
+        "        <default></default>"
+        "    </profiles>"
+        "    <users>"
+        "        <default>"
+        "            <password></password>"
+        "            <networks>"
+        "                <ip>::/0</ip>"
+        "            </networks>"
+        "            <profile>default</profile>"
+        "            <quota>default</quota>"
+        "        </default>"
+        "    </users>"
+        "    <quotas>"
+        "        <default></default>"
+        "    </quotas>"
+        "</yandex>";
+
+    ConfigurationPtr users_config;
+
+    if (config().has("users_config") || config().has("config-file") || fs::exists("config.xml"))
+    {
+        const auto users_config_path = config().getString("users_config", config().getString("config-file", "config.xml"));
+        ConfigProcessor config_processor(users_config_path);
+        const auto loaded_config = config_processor.loadConfig();
+        config_processor.savePreprocessedConfig(loaded_config, config().getString("path", DBMS_DEFAULT_PATH));
+        users_config = loaded_config.configuration;
+    }
+    else
+    {
+        users_config = getConfigurationFromXMLString(minimal_default_user_xml);
+    }
+
+    if (users_config)
+        global_context->setUsersConfig(users_config);
+    else
+        throw Exception("Can't load config for users", ErrorCodes::CANNOT_LOAD_CONFIG);
+}
+
+
+bool LocalServer::processMultiQuery(const String & all_queries_text)
+{
+    auto process_single_query = [&](const String & query_to_execute, const String &, ASTPtr)
+    {
+        try
+        {
+            processSingleQueryImpl(query_to_execute, query_to_execute, nullptr, echo_queries, false);
+        }
+        catch (...)
+        {
+            local_server_exception = std::make_unique<Exception>(getCurrentExceptionMessage(false), getCurrentExceptionCode());
+            have_error = true;
+        }
+    };
+
+    return processMultiQueryImpl(all_queries_text, process_single_query);
+}
+
+
+void LocalServer::processSingleQuery(const String & full_query)
+{
+    ASTPtr parsed_query;
+    if (is_interactive)
+    {
+        const auto * this_query_begin = full_query.data();
+        parsed_query = parseQuery(this_query_begin, full_query.data() + full_query.size(), false);
+    }
+
```

Comment:
Yes, because we parse query there in order to have nice echo query printing - done in parseQuery():
https://github.com/ClickHouse/ClickHouse/blob/9e990ea8ce40fd668a9c1703fa2f86a4b863affa/src/Client/ClientBase.cpp#L81-L88